swarm_dial.go

```diff
@@ -281,39 +286,311 @@ func (s *Swarm) dialPeer(ctx context.Context, p peer.ID) (*Conn, error) {
+		err      error
+		requests []int
+	}
+
+	reqno := 0
+	requests := make(map[int]*pendRequest)
+	pending := make(map[ma.Multiaddr]*addrDial)
+
+	var triggerDial <-chan time.Time
+	var nextDial []ma.Multiaddr
+	active := 0
+	done := false
+
+	resch := make(chan dialComplete)
+
+loop:
+	for {
+		select {
+		case req, ok := <-reqch:
+			if !ok {
+				// request channel has been closed, wait for pending dials to complete
+				if active > 0 {
+					done = true
+					reqch = nil
+					triggerDial = nil
+					continue loop
+				}
+
+				// no active dials, we are done
+				return
+			}
+
+			c := s.bestAcceptableConnToPeer(req.Ctx, p)
+			if c != nil {
+				req.Resch <- DialResponse{Conn: c}
+				continue loop
+			}
+
+			addrs, err := s.addrsForDial(req.Ctx, p)
+			if err != nil {
+				req.Resch <- DialResponse{Err: err}
+				continue loop
+			}
+
+			// at this point, len(addrs) > 0 or else it would be error from addrsForDial
+			// ranke them to process in order
+			addrs = s.rankAddrs(addrs)
+
+			// create the pending request object
+			pr := &pendRequest{
+				req:   req,
+				err:   &DialError{Peer: p},
+				addrs: make(map[ma.Multiaddr]struct{}),
+			}
+			for _, a := range addrs {
+				pr.addrs[a] = struct{}{}
+			}
+
+			// check if any of the addrs has been successfully dialed and accumulate
+			// errors from complete dials while collecting new addrs to dial/join
+			var todial []ma.Multiaddr
+			var tojoin []*addrDial
+
+			for _, a := range addrs {
+				ad, ok := pending[a]
+				if !ok {
+					todial = append(todial, a)
+					continue
+				}
+
+				if ad.conn != nil {
+					// dial to this addr was successful, complete the request
+					req.Resch <- DialResponse{Conn: ad.conn}
+					continue loop
+				}
+
+				if ad.err != nil {
+					// dial to this addr errored, accumulate the error
+					pr.err.recordErr(a, ad.err)
+					delete(pr.addrs, a)
+				}
+
+				// dial is still pending, add to the join list
+				tojoin = append(tojoin, ad)
+			}
+
+			if len(todial) == 0 && len(tojoin) == 0 {
+				// all request applicable addrs have been dialed, we must have errored
+				req.Resch <- DialResponse{Err: pr.err}
+				continue loop
+			}
+
+			// the request has some pending or new dials, track it and schedule new dials
+			reqno++
+			requests[reqno] = pr
+
+			for _, ad := range tojoin {
+				ad.requests = append(ad.requests, reqno)
+			}
+
+			if len(todial) > 0 {
+				for _, a := range todial {
+					pending[a] = &addrDial{ctx: req.Ctx, requests: []int{reqno}}
+				}
+
+				nextDial = append(nextDial, todial...)
+				nextDial = s.rankAddrs(nextDial)
+
+				if triggerDial == nil {
+					trigger := make(chan time.Time)
+					close(trigger)
+					triggerDial = trigger
+				}
+			}
+
+		case <-triggerDial:
+			if len(nextDial) == 0 {
+				triggerDial = nil
+				continue loop
+			}
+
+			next := nextDial[0]
+			nextDial = nextDial[1:]
+
+			// spawn the next dial
+			ad := pending[next]
+			go s.dialNextAddr(ad.ctx, p, next, resch)
```

Comment:
Please don't do this in a goroutine. We've had problems spawning too many goroutines here so we very carefully avoid it.

Reply:
I.e., we can get a bit backed up on concurrent dials.

Reply:
ok.

Reply:
done.